Qt: Reflect correct dbcache value in options panel.

valentinewallace opened the pull request:
Fixes: #11788
Shows correct value according to command line override or value in bitcoin.conf.

laanwj commented:
I don't think this is correct. The command line option should override the GUI setting with respect to the value used, but it should not set/change the GUI setting. Command line arguments shouldn't stick, which I think is what this does?

(reason is, for example: to be able to run with `-dbcache=10000` temporarily, then quit, then run without `-dbcache` and go back to the normal setting)

Imagallon909 commented:
On Dec 15, 2017 11:59 PM, "Wladimir J. van der Laan" <
notifications@github.com> wrote:

I don't think this is correct. The command line option should override the
GUI setting with respect to the value used, but it should not set/change
the GUI setting. Command line arguments shouldn't stick, which I think is
what this does?

(reason is, for example: to be able to run with -dbcache=10000 temporarily,
then quit, then run without -dbcache and go back to the normal setting)

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitcoin/bitcoin/pull/11915#issuecomment-352168178>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AgnKVt1Ns643WL33wn301hrIDpVdvJsUks5tA3hMgaJpZM4REQu3>
.

valentinewallace commented:
@laanwj Sorry I'm not sure I understand. When I run the GUI with this code, if I run with `-dbcache=10000` it will display 10000 in the options menu. If I then run without the flag, I see the either number in `bitcoin.conf` or the default in the options menu. 

Are you saying when you run without the flag a second time it still displays the last flag value used?

laanwj commented:
Yes. This:
```
settings.setValue("nDatabaseCache", (qint64)gArgs.GetArg("-dbcache", nDefaultDbCache));
```
Writes the value from the command line / bitcoin.conf to the QSettings object. That's not how it's supposed to work.

jonasschnelli commented:
To fix #11788, I would propose to show the "underlaying" dbcache value (set via bitcoin.conf or `-dbcache=<value>`) in the GUI settings pannel...

but only, if the value has not been overwritten by the QT settings (` (!settings.contains("nDatabaseCache")`).

Also, only write the GUI set database cache to QSettings when...
* Either `nDatabaseCache` is set
* Or the new value is not equal to the underlaying `-dbcache` value.

valentinewallace commented:
@jonasschnelli Just a slight clarification.

> but only, if the value has not been overwritten by the QT settings ((!settings.contains("nDatabaseCache")).

Due to `init` in optionsmodel, QSettings will always contain `nDatabaseCache`, it'll just contain the default if the user doesn't give it a new value through the GUI. However, this could be altered so it doesn't get set in `settings` unless the user sets it. 

Intuitively, my takeaway from your comment is that the GUI settings panel should display: 
- QSettings value if set by the user
- else flag value if set 
- else bitcoin.conf if set 
- else default value

And we only write to QSettings if a new value is set in the GUI, and never with the `-dbcache` flag value or bitcoin.conf.

jonasschnelli commented:
Oh. Your right...
Current code:
```
if (!settings.contains("nDatabaseCache"))
        settings.setValue("nDatabaseCache", (qint64)nDefaultDbCache);
```
This seems suboptimal. After executing this, you have no way to find out if the user has set a value or if we just have copied over the default from `nDefaultDbCache`.

I think it should only write a value to `nDatabaseCache` if it was set by the user and if it was different as the default.
Also, if `nDatabaseCache` is set, there could be a different color indicating that the value has been "overwritten".

But I agree, it should always show the current active database cache size.

valentinewallace commented:
That line you mentioned^ was there before I touched the file so I put it back in the spirit of making as few changes as possible to solve the problem (maybe I should've done the commits differently?), but I agree it does seem suboptimal. 

> I think it should only write a value to `nDatabaseCache` if it was set by the user and if it was different as the default.

Might it make sense to write a value even if the user sets it to the default value? Otherwise the bitcoin.conf value will override it even if the value the user wants in QSettings happens to be the default. 

Re: only writing a value to `nDatabaseCache` if user-set-- one tradeoff is that it makes the code for `init`ing the DatabaseCache value in QSettings different from how all the other QSettings values are initialized which is currently very uniform, adding a bit of complexity.   

However, I'll look into making that change and the color-coding tmrw :)

valentinewallace commented:
Okay, now we only write to QSettings if the user sets a non-default value!

jonasschnelli commented on src/qt/optionsmodel.cpp:
```diff
@@ -404,7 +406,7 @@ bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, in
             break;
         case DatabaseCache:
-            if (settings.value("nDatabaseCache") != value) {
+            if (settings.value("nDatabaseCache") != value && QVariant::fromValue(gArgs.GetArg("-dbcache", nDefaultDbCache)) != value && value != QVariant::fromValue(nDefaultDbCache)) {
```
What if someone has `dbcache=2000` in `bitcoin.conf` and then tries to reduce the dbcache via GUI back to 450?

valentinewallace replied:
Currently that wouldn't do anything because we don't write to QSettings if the user sets the GUI dbcache value to the default. Should we switch such that we write to QSettings no matter what value the user sets?

jonasschnelli commented:
Needs rebase

valentinewallace commented:
@jonasschnelli rebased!

Empact commented on src/qt/optionsmodel.cpp:
```diff
@@ -88,9 +88,10 @@ void OptionsModel::Init(bool resetSettings)
-    if (!m_node.softSetArg("-dbcache", settings.value("nDatabaseCache").toString().toStdString()))
+   std::string nDbCacheSettings = settings.contains("nDatabaseCache") ?
+     settings.value("nDatabaseCache").toString().toStdString() :
+     std::to_string(nDefaultDbCache);
```
whitespace issue here, you can use `git diff -U0 HEAD~3.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v` to fix: https://github.com/bitcoin/bitcoin/blob/master/contrib/devtools/README.md#clang-format-diffpy

valentinewallace replied:
good one, thanks!

Empact commented on src/qt/optionsmodel.cpp:
```diff
@@ -282,7 +283,11 @@ QVariant OptionsModel::data(const QModelIndex & index, int role) const
             return fCoinControlFeatures;
         case DatabaseCache:
-            return settings.value("nDatabaseCache");
+            if (!gArgs.GetArg("-dbcache", "").empty())
```
`IsArgSet` could be more explicit

valentinewallace replied:
fixed :)

Empact commented on src/qt/optionsmodel.cpp:
```diff
@@ -282,7 +283,11 @@ QVariant OptionsModel::data(const QModelIndex & index, int role) const
+                return (qint64)gArgs.GetArg("-dbcache", nDefaultDbCache);
+            if (settings.contains("nDatabaseCache"))
+                return settings.value("nDatabaseCache");
+            return QVariant::fromValue(nDefaultDbCache);
```
Seems you can return the default directly, the QVariant will be constructed from it.

valentinewallace replied:
I seem to be getting this error when I try to do that:

![deepinscreenshot_select-area_20181021212834](https://user-images.githubusercontent.com/10647624/47311351-56344900-d5ee-11e8-9d96-192a12514d2a.png)

Empact replied:
How about casting instead as above?

valentinewallace replied:
that works, and makes the code prettier! fixed :)

Empact commented on src/qt/optionsmodel.cpp:
```diff
@@ -282,7 +283,11 @@ QVariant OptionsModel::data(const QModelIndex & index, int role) const
         case DatabaseCache:
-            return settings.value("nDatabaseCache");
+            if (!gArgs.GetArg("-dbcache", "").empty())
+                return (qint64)gArgs.GetArg("-dbcache", nDefaultDbCache);
```
Why is this cast necessary?

valentinewallace replied:
Hm, for the code to compile? Am I missing something?

![deepinscreenshot_select-area_20181021215649](https://user-images.githubusercontent.com/10647624/47311377-6b10dc80-d5ee-11e8-9e39-2b71aeaafc50.png)

fanquake commented:
@valentinewallace Are you still interested in working on this? If so, could you address any nits & rebase. Otherwise we could close & label "Up for Grabs" etc.

valentinewallace commented:
@fanquake yes, sorry about that. I was planning to update it tomorrow if that's okay!

valentinewallace committed:
Qt: Reflect and persist correct dbcache value from options panel.

Resulting behavior:
- If the `dbcache` flag is set, the value is displayed but not persisted.
- If the user sets a value in the gui, it is persisted to QSettings.
- Else the default is displayed (but not persisted).

dongcarl commented:
No longer needed as #11788 is closed.
